Per-repeater HTTP codes for back-off

kaapstorm opened the pull request:
## Technical Summary

This change follows up [this PR comment](https://github.com/dimagi/commcare-hq/pull/36841#issuecomment-3185354635) by @millerdev 

## Safety Assurance

### Safety story

Tested locally.

### Automated test coverage

Yes

### QA Plan

No

### Migrations

- [x] The migrations in this code can be safely applied first independently of the code

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

kaapstorm committed:
Add fields to incl/excl HTTP_STATUS_BACK_OFF codes

kaapstorm committed:
Modify set_repeater_workers command

kaapstorm committed:
Rename management command for managing a repeater

kaapstorm committed:
repeater management command manages backoff codes

kaapstorm committed:
Check server failures with repeater backoff codes

kaapstorm committed:
Drop `is_traefik_proxy_404()`

kaapstorm committed:
Move code to Repeater, add tests

millerdev reviewed (commented):
Was the title of this PR intended to be `Per-repeater HTTP codes for back-off`?

kaapstorm commented on corehq/motech/repeaters/models.py:
```diff
@@ -299,6 +301,18 @@ class Repeater(RepeaterSuperProxy):
         default=REQUEST_POST,
         max_length=16,
     )
+    incl_backoff_codes = ArrayField(
+        base_field=models.IntegerField(),
+        default=list,
+        help_text=_('Also back off / retry these HTTP response codes'),
+        size=None,
+    )
+    excl_backoff_codes = ArrayField(
```
Is the objective to have a single field instead of two? I think using positive and negative integers would simplify the field(s), but would make the code less readable.

The other alternative I considered was to have a single field that defaults to `None`, and when you add or remove a code the field is set to a copy of `HTTP_STATUS_BACK_OFF` with that code added/removed. I chose against this because it would use more space than two fields, but maybe that's not really a concern. What do you think?

millerdev replied:
I think the extra logic would be trivial to implement, and it would mean the code updating the second field (`excl_backoff_codes`) could go away. And it would use slightly less storage, although that almost certainly does not matter since there are very few repeater records. Maybe not worth it, I'll let it up to you to decide.

kaapstorm replied:
:+1: 16dc8f137bb

kaapstorm committed:
Accurate help text, clearer output

Shows the end result in a helpful format. No longer shows underlying
field values. e.g.

```
$ python3 manage.py repeater demo b97c1b0992b742a980dd1c21bed863e3 \
    --set-max-workers 5 \
    --add-backoff-code 404

max_workers: 5
backoff_codes:
    404 "Not Found"
    408 "Request Timeout"
    423 "Locked"
    425 "Too Early"
    500 "Internal Server Error"
    502 "Bad Gateway"
    503 "Service Unavailable"
    504 "Gateway Timeout"
```

kaapstorm commented on corehq/motech/repeaters/management/commands/repeater.py:
```diff
@@ -0,0 +1,87 @@
+    def _get_repeater_info(self, repeater):
+        self.stdout.write(f'max_workers: {repeater.max_workers}')
+        self.stdout.write(f'incl_backoff_codes: {repeater.incl_backoff_codes}')
+        self.stdout.write(f'excl_backoff_codes: {repeater.excl_backoff_codes}')
```
84e4a845192

After adding `repeater.backoff_codes`, these two lines looked like they were making the output more complicated than it needed to be, so I removed them. In line with your comment about the help text, it's the end result that's relevant.

kaapstorm commented on corehq/motech/repeaters/models.py:
```diff
@@ -421,6 +435,33 @@ def num_workers(self):
+        return codes - set(self.excl_backoff_codes)
+
+    def add_backoff_code(self, code):
+        if code in self.backoff_codes:
```
Yes, I think so. I think the extra check would make this less readable, and would probably deserve a comment to explain why it's there, and the chances of it ever being required is really small. More likely is that a user will run `--add-backoff-code`, see that it's still not there, and run it again. ... But! Well spotted.

kaapstorm committed:
Use a single field for extra status codes

kaapstorm commented on corehq/motech/repeaters/models.py:
```diff
@@ -421,6 +427,43 @@ def num_workers(self):
+            if code > 0:
+                codes.add(code)
+            else:
+                # Multiply by -1 to get the HTTP status code.
+                codes.remove(code * -1)
```
Huh! I never knew `-foo` was a thing!

kaapstorm committed:
Negate with `-`, use `discard()`
